src/Adapter/MSTest.TestAdapter/Helpers/ReflectHelper.cs

```diff
@@ -516,102 +379,74 @@ internal virtual IEnumerable<Trait> GetTestPropertiesAsTraits(MemberInfo testPro
-    /// <param name="type">The type to inspect.</param>
-    /// <param name="inherit">Look at inheritance chain.</param>
-    /// <returns>An instance of the attribute.</returns>
-    internal static TAttributeType? GetDerivedAttribute<TAttributeType>(Type type, bool inherit)
-        where TAttributeType : Attribute
-    {
-        object[] attributes = GetCustomAttributes(type.GetTypeInfo(), inherit);
-        if (attributes == null)
-        {
-            return null;
-        }
-
-        // Try to find the attribute that is derived from baseAttrType.
-        foreach (object attribute in attributes)
-        {
-            DebugEx.Assert(attribute != null, "ReflectHelper.DefinesAttributeDerivedFrom: internal error: wrong value in the attributes dictionary.");
-
-            Type attributeType = attribute.GetType();
-            if (attributeType.Equals(typeof(TAttributeType)) || attributeType.IsSubclassOf(typeof(TAttributeType)))
-            {
-                return attribute as TAttributeType;
-            }
-        }
-
-        return null;
     }
 
     /// <summary>
     /// Returns owner if attribute is applied to TestMethod, else null.
     /// </summary>
     /// <param name="ownerAttributeProvider">The member to inspect.</param>
     /// <returns>owner if attribute is applied to TestMethod, else null.</returns>
-    private static string? GetOwner(MemberInfo ownerAttributeProvider)
+    private string? GetOwner(MemberInfo ownerAttributeProvider)
     {
-        OwnerAttribute[] ownerAttribute = GetCustomAttributes<OwnerAttribute>(ownerAttributeProvider, true);
+        OwnerAttribute? ownerAttribute = GetFirstDerivedAttributeOrDefault<OwnerAttribute>(ownerAttributeProvider, inherit: true);
 
-        return ownerAttribute.Length != 1
-            ? null
-            : ownerAttribute[0].Owner;
+        return ownerAttribute?.Owner;
     }
 
     /// <summary>
-    /// Get the Attributes (TypeName/TypeObject) for a given member.
+    /// Gets and caches the attributes for the given type, or method.
     /// </summary>
-    /// <param name="memberInfo">The member to inspect.</param>
+    /// <param name="attributeProvider">The member to inspect.</param>
     /// <param name="inherit">Look at inheritance chain.</param>
     /// <returns>attributes defined.</returns>
-    private Dictionary<string, object>? GetAttributes(MemberInfo memberInfo, bool inherit)
+    private Attribute[] GetCustomAttributesCached(ICustomAttributeProvider attributeProvider, bool inherit)
     {
+        if (inherit)
+        {
+            lock (_inheritedAttributeCache)
+            {
+                return GetOrAddAttributes(_inheritedAttributeCache, attributeProvider, inherit: true);
+            }
+        }
+        else
+        {
+            lock (_nonInheritedAttributeCache)
+            {
+                return GetOrAddAttributes(_nonInheritedAttributeCache, attributeProvider, inherit: false);
+            }
+        }
```

Comment:
What about using `ConcurrentDictionary` instead?

Reply:
I was thinking about it I wanted to measure the impact once this is merged, I saw time spent in locks in PerfView, but it also was not significantly faster without any locks. So please merge this approach first.

Reply:
I was also thinking about writer only lock, but I am not sure if concurrent dictionary does not do that for me automatically already.

Reply:
https://github.com/microsoft/testfx/issues/3042